### ISSUE TITLE
feat: add `rename_if_exists` or `diff_if_exists` support to configuration

### ISSUE THREAD

**Issue by brutus** [copier-org/copier#1262]
### Actual Situation

I usually provide some default configuration files for domain-specific stuff in my templates. That works fine, when I use the template for fresh projects, but tends to clash with already existing files, when I switch to using templates later in existing projects.

While `skip_if_exists` prevents overwriting project files, I often lose some "newer" settings I was tracking in my template but was not already using in the project - which I then had to diff from the template and add to the project file manually.

Since I use templates not only for scaffolding projects but also a lot - probably more often - for keeping basic config files (e.g. editorconfig, pre-commit, etc.) in a central place, trying to keep them updated in all my projects, I tend to run into this.

### Desired Situation

I tend to work around this with a "namespaced" version of the file, e.g. `pyproject:mytemplate.toml`. Using `tasks:` to "un-namespace" it, if not already present in the project. 

While this has the tendency to result in some "boilerplate" (especially if I went wild and used `if ! grep -F 'tool.ruff' pyproject.toml; then echo | cat pyproject.toml - pyproject:mytemplate:ruff.toml > .tf; mv .tf pyproject.toml; fi`, or similar a couple of times…), it generally worked okay. Since `tasks` are now hidden behind a flag (which is great IMO), I try to rely less on them though.

Basically, I'm looking for a way to preserve the difference when overwriting files on _copier copy_ runs, which requires less boilerplate and leans more on the declarative side.

### Proposed solution

Overwriting on copy and using diff to get the desired output is possible to be fair. I was just looking for a less manual solution :wink: 

Not quite sure how to tackle this though… I was thinking maybe a feature similar to [skip_if_exists](https://copier.readthedocs.io/en/stable/configuring/#skip_if_exists), but that renames the file, if it already exists, or adds a diff/patch or "merge—markers" in the file instead of overwriting?

**Comment by yajo** [comment]
Do you really think that these options would add value? 🤔

When you copy the 1st time and copier overwrites pyproject.toml, you
already get a diff just with git. How would be better a diff generated by
copier? It's gonna be the same diff after all...

BTW, to enable features in that file, the "appropriate" way would be to ask
questions, not to read the pre-existing file and add portions in a task.

In my case I tend to configure tools in separate files. It's uglier than
having all in a single pyproject.toml file, but it's easier to maintain
with copier.

**Comment by brutus** [comment]
> Do you really think that these options would add value? :thinking: 

Well, I tried to get my doubts across and formulate more like "looking for perspective" than a straight-up feature request. That said, yes, I think that `rename_if_exists` _could_ add value, the rest was just me spitballing to give some context. Anyway, does it add enough value to justify the inclusion, IDK. Of course, feel free to close this.

I just want to add some more thoughts thought (since I feel my example might have come across in a nonintended way - I choose `pyproject.toml` b/c it's well known, which might lead away from my actual concern):

I know that _the "appropriate" way would be to ask questions_. I do that - it's how the file and its content are generated. Sticking with my example: In a template for _pre-commit_ I might ask if the project uses Python and if so, _ruff_ and _black_ are included as linters in the generated pre-commit config (I do not ask which specific linters to use, b/c that might change over time and should stay consistent across all projects and governed by the template). They might need some configuration though, b/c I want to encourage a common baseline… and this might clash with the already existing configuration for the project. But I do not just want to `skip_if_exists`, I like the new stuff to be present, without overwriting the old :thinking: 

Aside: While I too personally tend to prefer to _configure tools in separate files_, that's not always possible (e.g. it would work for _ruff_, true, but not for _black_ IIRC).

El mié., 19 jul. 2023 16:35, Brutus ***@***.***> escribió:
> I'm looking for a way to preserve the difference when overwriting files on *copier copy* runs, which requires less boilerplate and leans more on the declarative side. Overwriting on copy and using diff to get the desired output is possible to be fair.

I might have an uncommon use case: while I write the template, I'm usually not the one who applies them to existing projects in bulk. Those who do generally have a good understanding of their project, but not necessarily about the meta stuff the template is adding. It's quite easy to lose a line in a diff etc. I was just looking for a way to make this process easier.

Anyway, thank you, for taking the time to respond. And I appreciate not giving in to "feature bloat", it's a pretty specific use case. But maybe someone stumbles over this, who has a neat solution at hand, I overlooked :)

**Comment by yajo** [comment]
Well, honestly this really does seem like a big corner case.

The bottom reason why you really seem to need this is mostly that the person who applies the template (only the 1st time, of course) doesn't really know what he's doing... 🤔 

If that were the case in my company, I'd say the solution is training. A README, a video or similar, where people know how to proceed. Maybe even a script they can run.

Regarding rename, you rename it if it's found. But then, what to do if the rename is found too? And aside from being a strange workaround for this use case, what other uses would this have? I really can't seem to find this feature fitting well in Copier's product vision.

Regarding the diff... I'm not even sure where to start doing this in a way that serves a more global purpose.

FWIW remember that you can use copier as a library. Maybe you could add your own domain-specific tool that applies some logic on top of copier?

Another crazy idea: tell your users to apply the template like this:

```sh
copier copy $src $dst --data original_pyproject="$(cat pyproject.toml)"
```

With some imagination you can set that as a secret question, so it doesn't get recorded in the answers file. Then you can do some jinja-fu to find the important parts from that file, without having to bother the user too much.

However, in this case I guess the replay for updates won't be very accurate...

I hope some of these ideas serves you, but for the time being I think I'm gonna reject this proposal. Thanks anyway for reaching out. Remember that you also have the discussion forum. Feel free to use it to ask for help to other users, or to share your final solution!

Regards.
